### DISCUSSION TITLE
Adapter with Sensor data and Geo Positions Implementation

### DISCUSSION THREAD

**Discussion by flomickl** [apache/streampipes#1034] [Ideas]
Hi,

I found a geodata source for Open Sky Network’s live API

here is a request example for airplanes over Switzerland via bounding box:
https://opensky-network.org/api/states/all?lamin=45.8389&lomin=5.9962&lamax=47.8229&lomax=10.5226

Because there are no key value pairs, the adapters cannot fetch the data at the moment.
I was unable to do so.

The key values of the response are listed here. https://openskynetwork.github.io/opensky-api/rest.html#response

So, do you think it is possible to create an adapter where the key value pairs are created correctly?
Similar to the ISS Adapter but more advance because of the missing keys.

Greetings
Florian

===UPDATE===

in general: I added a page in the wiki with good data sources with geo positions. https://cwiki.apache.org/confluence/x/nYpbDg
I think OpenSenseMap is for example a good way to combine real sensor data with position data.

There is also a OGC standard called SensorThings API. Have you ever heard of it?

https://developers.sensorup.com/docs/#introduction
http://developers.sensorup.com/webinars/Introduction-to-SensorThings/

=== UPDATE 2 ====
 water level gauging station in Germany cut of in adapter
https://www.pegelonline.wsv.de/webservices/rest-api/v2/stations/BONN.json?includeTimeseries=true&includeCurrentMeasurement=true

**Comment by dominikriemer** [comment]
Hi Florian,
while this would certainly be a great adapter, unfortunately we can't directly add the adapter to StreamPipes due to license issues (API is free for research and non-commercial usage) [1].

We could however work on supporting the key value format and you could, for instance, create a specific adapter in your own repository and maybe we could also add references to "unsupported community-created extensions" to the website.

[1] https://openskynetwork.github.io/opensky-api/index.html

**Comment by flomickl** [comment]
@dominikriemer 
that sounds like a good idea.

What does it need to support the key value format?

**Reply by dominikriemer** [reply]
I think we just need to add a new format, @tenthe is currently working on improving data formats in StreamPipes Connect and we can discuss if we can include this there.

**Comment by flomickl** [comment]
@tenthe Do you think this is easy to do with the new, improved data formats?


@dominikriemer I did not realize the terms of use. But I am willing and happy to create an external adapter source, if compatible.
Is there already a workflow on how to add external adapters? 

in general: I added a page in the wiki with good data sources with geo positions. https://cwiki.apache.org/confluence/x/nYpbDg
I think OpenSenseMap  is for example a good way to combine real sensor data with position data.

There is also a OGC standard called SensorThings API. Have you ever heard of it?

https://developers.sensorup.com/docs/#introduction
http://developers.sensorup.com/webinars/Introduction-to-SensorThings/

What do you think?

**Reply by tenthe** [reply]
So far I have only prepared it to remove the `Data Set` models from the code. 
Next I will delete them. After that I would suggest to harmonize the Specific and Generic Adpater first.

Once that is completed we can look at the formats. 
But I guess this will take a little more time.

**Comment by flomickl** [comment]
btw, I tried the water level gauging station in Germany as source, but I can't access the measurement values in the adapter settings. It is kind of cut off.

https://www.pegelonline.wsv.de/webservices/rest-api/v2/stations/BONN.json?includeTimeseries=true&includeCurrentMeasurement=true

**Reply by bossenti** [reply]
Thats probably due to the nested structure of the JSON.
In this case you currently need to implement a dedicated adapter that preprocesses the JSON.
Am I right, @tenthe?

**Reply by tenthe** [reply]
Yes, @bossenti is right. 
The values of the sensor are in an array. This is currently not supported by the formats.

**Reply by flomickl** [reply]
@bossenti  ok, if I understand you correctly: it could be done in a separate adapter?
While parsing the JSON and referencing the output values to a specific output structure?

**Reply by bossenti** [reply]
Yes exactly :)

**Reply by flomickl** [reply]
so a little bit like the ISS Adapter but more advanced? :)

**Reply by bossenti** [reply]
Yes, I think so

**Comment by flomickl** [comment]
just to make it complete 
another source
also with a sensorthing

MQTT-Broker: iot.hamburg.de

 Topic:

- v1.0/Observations
- v1.0/Datastreams({id})/Observations

as json:
https://iot.hamburg.de/v1.0/Things?$filter=Datastreams/properties/serviceName%20eq%20%27HH_STA_TEC_Energiedaten_HH-Bergedorf%27&$count=true&$expand=Locations,Datastreams($expand=Observations($orderby=phenomenonTime%20desc;$top=3),Sensor,ObservedProperty)


Overwiev: https://metaver.de/trefferanzeige?docuuid=785D987C-AAFF-471D-AE3A-EBCD4C9E23F1#detail_links

**Comment by flomickl** [comment]
not a sensor dataset but the changeset of OSM

https://api.openstreetmap.org/api/0.6/changesets
as xml and tag changeset

it is working but the tags list can't be included so far.
